Improve tests for ROOT consistency

robbietuk opened the pull request:
Replaces the #23 PR because of the merge of `test_view_offset_root` https://github.com/UCL/STIR/pull/991 and https://github.com/UCL/STIR/pull/1016.

Adds test __`test_consistency_with_root`__. 
This PR combines `test_view_offset_root` (in non-TOF `UCL/STIR`) and the TOF `test_consistency_root` (in `tof_sino_ucl`) tests into __`test_consistency_with_root`__ . The conbination of these two tests is to reduce the number of list mode readings and allow for a comparison between the TOF and non-TOF functionality.

___

Current errors... 
```
Test Results
--------------------------------------------------------------
	Test Index	|	nonTOF	|	TOF
--------------------------------------------------------------
		1	|	Pass	|	Fail
		2	|	Pass	|	Fail
		3	|	Pass	|	Fail
		4	|	Pass	|	Fail
		5	|	Pass	|	Pass
		6	|	Pass	|	Fail
		7	|	Pass	|	Fail
		8	|	Pass	|	Fail

End of tests. Please correct errors !
```
The only test passing (5) is the point source at (0, 0, 145.267) (x,y,z), i.e. offset in z axis. This point source data may be considered almost 2D because it exists in the edge of the FOV of the scanner so only the last ring should contribute towards the data. This does concern me that there is an issue with TOF and segments issue somewhere... However `TOF_distance_threshold` might be too strict...

eliseemond committed:
Some changes for positron range: one medium

eliseemond committed:
Files for STIR+ROOT check (symmetries, unlisting etc.)

eliseemond committed:
The .hroot files were missing

eliseemond committed:
Gate/STIR time resolution calibration

For more information, please check the 2017 MIC presentation I have done
during the STIR user's meeting.

eliseemond committed:
Update README.md

eliseemond committed:
README.md for ROOT consistency check for TOF

robbietuk committed:
Move ROOT-STIR-consistency into recon_tests and enable test [ci skip]

robbietuk committed:
Remove consistency configuration and generation files

robbietuk committed:
Redesign of test_view_offset_root to include TOF tests

Also required a lot of method/variable renaming to make this logically work.

robbietuk committed:
Rename test_view_offset_root -> test_consistency_with_root

robbietuk committed:
Remove old test test_consistency_root

robbietuk commented on src/recon_test/test_consistency_with_root.cxx:
```diff
@@ -0,0 +1,547 @@
+
+  // Failure conditioner and recording
+  nonTOF_distance_threshold = 1.5 * norm(grid_spacing); // Using norm(grid_spacing) as a nonTOF_distance_threshold
+  TOF_distance_threshold = 1.5 * norm(grid_spacing); // Using norm(grid_spacing) as a tof_distance_threshold
```
Too harsh for TOF?

robbietuk commented on src/recon_test/test_consistency_with_root.cxx:
```diff
@@ -0,0 +1,547 @@
+ *  \author Elise Emond
+ *  \author Robert Twyman
+ *
+ * This test currently uses Root listmodes of single point sources.
+ * This test computes the distance between the original point source position and the closes voxel that the list mode
+ * event passes through. Ideally each event would travel directly through the original point source position but
+ * error may be present. Therefore we test that the majority of LORs travel close enough.
```
Needs updating for TOF

KrisThielemans replied:
@robbietuk  this still needs to be done

robbietuk commented on src/recon_test/test_consistency_with_root.cxx:
```diff
@@ -0,0 +1,547 @@
+  CartesianCoordinate3D<float> coord;
+  //! Mass of all weights
+  float value;
+  //! COM standard deviation.
```
```suggestion
```

robbietuk commented on src/recon_test/test_consistency_with_root.cxx:
```diff
@@ -0,0 +1,547 @@
+
+#include "stir/warning.h"
+
+#include <stack>
```
```suggestion
```

robbietuk commented on src/recon_test/test_consistency_with_root.cxx:
```diff
@@ -0,0 +1,547 @@
+      : coord(voxel_centre_v), value(value_v)
+  {}
+
+  //! Center of mass position in mm
+  CartesianCoordinate3D<float> coord;
+  //! Mass of all weights
+  float value;
```
```suggestion
  //! Coordinate position
  CartesianCoordinate3D<float> coord;
  //! Weight value
  float value;
```

robbietuk commented on src/recon_test/test_consistency_with_root.cxx:
```diff
@@ -0,0 +1,547 @@
+
+private:
+
+  void setup();
```
Add documentation for `setup()`

robbietuk commented on src/recon_test/test_consistency_with_root.cxx:
```diff
@@ -0,0 +1,547 @@
+   * along a bin LOR). This method should include all tests/ setup for additional test such that the list data is
+   * read only once.
+   *
+   * Passes ProjMatrixElemsForOneBin (LOR) to test_nonTOF_LOR_closest_approach() and if fails,
+   * add 1 to `failed_events` (LOR's closest voxel was not within nonTOF_distance_threshold).
+   * Check if the number of `failed_events` is greater than half the number of tested events to pass the test.
```
Update non-TOF documentation and document TOF test interaction

KrisThielemans replied:
@robbietuk still to do

robbietuk commented on src/recon_test/test_consistency_with_root.cxx:
```diff
@@ -0,0 +1,547 @@
+   */
+  void process_list_data();
+
+  void loop_through_list_events(shared_ptr<CListModeData> lm_data_sptr, const shared_ptr<ProjMatrixByBin>& proj_matrix_sptr,
+                                const CartesianCoordinate3D<float> original_coords);
```
Document

robbietuk commented on src/recon_test/test_consistency_with_root.cxx:
```diff
@@ -0,0 +1,547 @@
+
+  void loop_through_list_events(shared_ptr<CListModeData> lm_data_sptr, const shared_ptr<ProjMatrixByBin>& proj_matrix_sptr,
+                                const CartesianCoordinate3D<float> original_coords);
+  /*! Given a ProjMatrixElemsForOneBin (probabilities), test if the closest voxel in the LOR to the original_coords
+   * is within nonTOF_distance_threshold distance. If it is, pass with true, otherwise fales.
+   * @param probabilities ProjMatrixElemsForOneBin object of a list mode event
+   * @param test_discretised_density_sptr Density containing a point source.
+   * @param original_coords Precomputed coordinates of the point source
+   * @param tolerance Precomputed voxel sizes - used in setting nonTOF_distance_threshold
+   * @return True if test passes, false if failed.
+   */
```
```suggestion
  /*! Given a ProjMatrixElemsForOneBin (probabilities), test if the closest voxel in the LOR to the original_coords
   * is within nonTOF_distance_threshold distance. If it is, pass with true, otherwise fales.
   * @param probabilities ProjMatrixElemsForOneBin object of a list mode event
   * @return True if test passes, false if failed.
   */
```

robbietuk commented on src/recon_test/test_consistency_with_root.cxx:
```diff
@@ -0,0 +1,547 @@
+   * @return True if test passes, false if failed.
+   */
+  bool test_nonTOF_LOR_closest_approach(const ProjMatrixElemsForOneBin& probabilities);
+  void post_processing_nonTOF();
```
Document

robbietuk commented on src/recon_test/test_consistency_with_root.cxx:
```diff
@@ -0,0 +1,547 @@
+  void post_processing_nonTOF();
+
+
+  //! Selects and stores the highest probability elements of ProjMatrixElemsForOneBin.
+  bool test_TOF_max_lor_voxel(const ProjMatrixElemsForOneBin& probabilities);
```
Improve documentation

robbietuk commented on src/recon_test/test_consistency_with_root.cxx:
```diff
@@ -0,0 +1,547 @@
+  //! Selects and stores the highest probability elements of ProjMatrixElemsForOneBin.
+  bool test_TOF_max_lor_voxel(const ProjMatrixElemsForOneBin& probabilities);
+
+  //! Checks if original and calculated coordinates are close enough and saves data to file.
+  void post_processing_TOF();
```
Improve documentation

robbietuk commented on src/recon_test/test_consistency_with_root.cxx:
```diff
@@ -0,0 +1,547 @@
+  //! Checks if original and calculated coordinates are close enough and saves data to file.
+  void post_processing_TOF();
+
+  //! Modified version of check_if_equal for this test
+  bool check_if_almost_equal(const float a, const float b, const std::string str, const float tolerance);
```
Method is unused. Can remove

robbietuk commented on src/recon_test/test_consistency_with_root.cxx:
```diff
@@ -0,0 +1,547 @@
+
+  int test_index;
+  int num_tests;
+  std::vector<bool> test_results_nonTOF;
+  std::vector<bool> test_results_TOF;
```
```suggestion
  // Test results storage. Records if each test failed or not. E.g.`test_index` result is stored in `test_results_nonTOF[test_index -1]`
  std::vector<bool> test_results_nonTOF;
  std::vector<bool> test_results_TOF;
```

robbietuk commented on src/recon_test/test_consistency_with_root.cxx:
```diff
@@ -0,0 +1,547 @@
+ROOTConsistencyTests::
+setup()
+{
+  {
+    // Create the point source (discretised_density_sptr) with GenerateImage from parameter file
+    // GenerateImage requires `const char* const par_filename`.
+    GenerateImage image_gen_application(get_generate_image_par_filename().c_str());
+    image_gen_application.compute();
+    discretised_density_sptr = image_gen_application.get_output_sptr();
+  }
```
Remove `{ }`

robbietuk commented on src/recon_test/test_consistency_with_root.cxx:
```diff
@@ -0,0 +1,547 @@
+  nonTOF_closest_voxels_list.empty();
+  max_coords_for_each_event.empty();
+
+  // The number of LORs with closes approach greater than the threshold.
```
```suggestion
  // Reset the number of processed and failed events
```

robbietuk commented on src/recon_test/test_consistency_with_root.cxx:
```diff
@@ -0,0 +1,547 @@
+  /// TOF VARIABLES
+  //! A copy of the scanner timing resolution in mm
+  float gauss_sigma_in_mm;
+  std::vector<WeightedCoordinate> max_coords_for_each_event;
```
`max_coords_for_each_event` is a misleading variable name. Consider `LOR_peak_value_coords`? Peak is more descriptive and LOR is clearer than event.

robbietuk replied:
Add `TOF` to indicate TOF test variable -> `TOF_LOR_peak_value_coords`

robbietuk commented on src/recon_test/test_consistency_with_root.cxx:
```diff
@@ -0,0 +1,547 @@
+  cerr << "\nNon-TOF number of failed events: " << num_failed_nonTOF_lor_events
+       << "\tNumber of tested events: = " << num_events_tested << std::endl;
+
+  test_results_nonTOF[test_index] = check_if_less(num_failed_nonTOF_lor_events, failure_tolerance_nonTOF * num_events_tested,
```
WRONG! `Test index - 1`

robbietuk commented on src/recon_test/test_consistency_with_root.cxx:
```diff
@@ -0,0 +1,547 @@
+  cerr << "\nTOF number of failed events: " << num_failed_TOF_lor_events
+       << "\tNumber of tested events: = " << num_events_tested << std::endl;
+
+  test_results_TOF[test_index] = check_if_less(num_failed_TOF_lor_events, failure_tolerance_TOF * num_events_tested,
```
WRONG `test_index -1`

robbietuk commented on src/recon_test/test_consistency_with_root.cxx:
```diff
@@ -0,0 +1,547 @@
+    myfile.close();
+  }
+
+#if 0 // old code, will not run. Used for records but can be deleted...
```
Can be removed now.

robbietuk commented on src/recon_test/test_consistency_with_root.cxx:
```diff
@@ -0,0 +1,547 @@
+#endif
+}
+
+bool
+ROOTConsistencyTests::
+check_if_almost_equal(const float a, const float b, const std::string str, const float tolerance)
+{
+  if ((fabs(a-b) > tolerance))
+  {
+    std::cerr << "Error : unequal values are " << a << " and " << b
+              << ". " << str << std::endl;
+    everything_ok = false;
+    return false;
+  }
+  else
+    return true;
+}
```
Unused method

robbietuk reviewed (commented):
Self code review before GHA should be rerun.

robbietuk commented on src/recon_test/test_consistency_with_root.cxx:
```diff
@@ -0,0 +1,547 @@
+                                 discretised_density_sptr);
+
+  // TOF requirements
+  proj_matrix_sptr->enable_tof(lm_data_sptr->get_proj_data_info_sptr());
```
Build error
```
/home/runner/work/STIR/STIR/src/include/stir/recon_buildblock/ProjMatrixByBin.h:246:8: note: declared private here
  246 |   void enable_tof(const shared_ptr<const ProjDataInfo>& proj_data_info_sptr,const bool v = true);
      |        ^~~~~~~~~~
```

Locally I have moved `enable_tof` to a public method. Should we do this in the TOF branch?

robbietuk replied:
In the latest change, the `enable_tof()` call was removed.

robbietuk commented:
Increasing the `TOF_distance_threshold = 2.5 * norm(grid_spacing);` (i.e. 1.5 -> 2.5) results in the following results

```
Test Results
-------------------------------------------------------
	Test Index	|	nonTOF	|	TOF
-------------------------------------------------------
		1	|	Pass	|	Fail
		2	|	Pass	|	Fail
		3	|	Pass	|	Fail
		4	|	Pass	|	Fail
		5	|	Pass	|	Pass
		6	|	Pass	|	Pass
		7	|	Pass	|	Pass
		8	|	Pass	|	Pass

End of tests. Please correct errors !
```

Tests 5,6,7,8 are all z-shifted tests. Coincidences for these datasets likely occur in segment 0 viewgrams, e.g. 2D. This suggests that there is a segment issue with TOF? I need to python debugging code ready to conclude that though.

robbietuk committed:
[ci skip] Significant fixes for test_consistency_with_root tests

- Updates some documentation, 
- removes `check_if_almost_equal` method because unused,  
- renames `masx_coords_for_each_event` to `TOF_LOR_peak_value_coords`,
-fix issue with `num_test` and `num_tests` being different variables,
- fix issue with `empty()` being used to clear a std::vector (correct method is `clear()`,
- increase `TOF_distance_threshold` to 2.5*norm(grid_spacing),
- fix indexing issue in `test_results_nonTOF` and `test_results_TOF`, and
- remove unused code block

robbietuk committed:
[ci skip] The data is TOF by default and guass_sigma_in_mm is not needed for test.

robbietuk committed:
Rename debug tool and cleanup content for non-TOF tests

robbietuk committed:
Update test data output ti *voxel_data*.txt in .gitignore

robbietuk committed:
[ci skip] TOF debug_consistency_with_root improvements

Adds PointCloud3D() and changes filename checks?

robbietuk commented:
TODO
----

- [ ] Incraese TOF tolerance. The plots produced by the debug indicate the data is approximately in the correct location.

There may be an issue with large ring difference LORs causing problems. This might be due to the inconsitency between STIR and GATE geometries in z axis as currently we are not really testing for this. 

- [ ] Investigate the cause of the zero z shifted (center of the scanner) events having a lower pass rate?
- [ ] Investigate if that problem is a result of the non-TOF case.
- [ ] Increase the TOF resolution in GATE and STIR to see if that removes some of the errors 

----
Possible methods
----

- [ ] C++ test: Output all voxel coords (and values) for each event **OR** output the LOR vectors (with some value information). 
- [ ] Python: Visualise if the LOR is _near_ the origin for these outlier events. 
- [ ] Investigate other causes of these outliers.

NicoleJurjew commented:
I'm happy to re-run the GATE simulations on the cluster with increased TOF-rate. As I see in the GATE - macros, currently they are run with 10 ps timing resolution. Which resolution do you suggest to run them with?

Edit: Sorry, below solution is no option because the settings were such that only coincidences were saved in the root files, not the hits.
[There is a way to use the root-files we currently have and re-sort the events with different digitizer settings (it's called DigiGATE, see here: https://opengate.readthedocs.io/en/latest/how_to_run_gate.html?highlight=DigiGATE#how-to-launch-digigate ). I haven't done this yet, and I don't know the computation time of the source generation, but I can try this out as well if you'd prefer a "direct" comparison.]

robbietuk commented:
The GATE simulations are very cheap to recompute and take less than 10 minutes. There are approxiamtely 10^3 or 10^4 events. The hits are not saved as they are relatively expensive to store and file size needed to be small.

NicoleJurjew commented:
I ran GATE - simulations with the exact same source-positions as in the zenodo files, just with different timing resolutions in the digitiser.mac.

TOF-resolution 10 ps (as data on zenodo):
![tol25_TOF10](https://user-images.githubusercontent.com/86268422/176934847-0e9b916d-1bf1-4876-92dc-5d948d8a117d.png)

TOF-resolution 0 ps:
![tol25_TOF0](https://user-images.githubusercontent.com/86268422/176934900-7fe1cc4d-5954-4092-a3a4-4e883ed7f27f.png)

TOF-resolution 500 ps:
![tol25_TOF500](https://user-images.githubusercontent.com/86268422/176934927-c312e91d-873e-46ad-b48c-8cba592d59f7.png)

I find the results for 500 ps very surprising as for the last three sources, there is again not a single failed event:
![image](https://user-images.githubusercontent.com/86268422/176935059-eba1f2e0-ca9d-4afc-b529-74d95d88f2d5.png)

There is one thing that I noticed: in the root-header files on zenodo (and in the ROOT_STIR_consistency folder in the SuperBuild), the timing resolution was set to 75 ps, while the GATE macros state the digitiser was set to 10 ps. Doesn't this value (and the TOF-bin size and number with it) need to be adapted?

NikEfth commented:
>  the timing resolution was set to 75 ps, while the GATE macros state the digitizer was set to 10 ps

It is the effect of the DOI. You cannot go lower that the uncertainty in the detection position in the crystal. 
All three papers I have in TOF, at one point or another discuss this.

robbietuk commented:
Great job! Decreasing crystal size should help somewhat but it isn't necessary imo. I think increase the tolerance so the tests pass now.

NicoleJurjew committed:
changed TOF-test threshold to 3D & added 3D plot of all sources in 1 plot

NicoleJurjew committed:
adapted TOF-test threshold to 3.3

NicoleJurjew committed:
add documentation and changee variable names and plot legend

robbietuk committed:
Merge pull request #26 from NicoleJurjew/adapt_root_consistency_threshold

Adapt root consistency threshold

robbietuk commented:
@KrisThielemans I think this is ready now with @NicoleJurjew's changes. It is only failing on the `test_time_of_flight`, which is expected right now.

KrisThielemans commented on examples/ROOT_files/ROOT_STIR_consistency/DebugScripts/debug_consistency_with_root.py:
```diff
@@ -73,20 +73,116 @@ def plot_1D_distance_histogram(distances, n_bins=100, logy=False, pretitle=""):
     """
 
-    def __init__(self, filename, tolerance=6.66983):
+    def __init__(self, filename, tolerance=1.5 * 4.447):
```
what are these numbers? Can it be documented?

robbietuk replied:
Yes, it is information now included in the CTest output files

KrisThielemans commented on src/recon_test/test_consistency_with_root.cxx:
```diff
@@ -0,0 +1,402 @@
+
+  /// TOF VARIABLES
+  //! A copy of the scanner timing resolution in mm
+  const float failure_tolerance_TOF = 0.05;
```
hardwired. can we get this from the data? The name doesn't match the doc.

robbietuk replied:
We no longer use scanner timing resolution, I have removed.

KrisThielemans commented on src/recon_test/test_consistency_with_root.cxx:
```diff
@@ -0,0 +1,402 @@
+
+  /// NON TOF VARIABLES
+  //! A vector to store the coordinates of the closest approach voxels.
+  const float failure_tolerance_nonTOF = 0.05;
```
this variable needs to be moved before the doxygen comment, but it also needs its own comment

robbietuk replied:
Done

KrisThielemans commented on src/recon_test/test_consistency_with_root.cxx:
```diff
@@ -0,0 +1,402 @@
+
+  // Failure conditioner and recording
+  nonTOF_distance_threshold = 1.5 * norm(grid_spacing); // Using norm(grid_spacing) as a nonTOF_distance_threshold
+  TOF_distance_threshold = 3.3 * norm(grid_spacing); // With the value 3.3*norm(grid_spacing), test_root_consistency passes for all sources
```
This threshold has to depend on the TOF resolution. Computing it is hard, but we can now work our way backwards. Something like this:
```suggestion
  // set TOF distance threshold
  {
    // With the default files, we found that 3.3*norm(grid_spacing) is a reasonable limit.
    // We try to generalise this to other data (although it will likely need further
    // adjustment for cases with very different TOF resolution).
    const auto& pdi = *lm_data_sptr->get_proj_data_info_sptr();
    const float approx_tof_resolution =
      pdi.get_scanner()*get_timing_resolution()*pdf.get_tof_mash_factor();
    const float org_threshold = 3.3F / 75; // CHECK
    this->TOF_distance_threshold = approx_tof_resolution * org_threshold * norm(grid_spacing);
  }
```
or should `failure_tolerance_TOF` be used (according to its docstring, but I doubt it)

KrisThielemans replied:
obviously, at this point, we don't have `lm_data_sptr`, so this (or the `proj_data_info_sptr) will need to be passed to `set_up()`

robbietuk replied:
`get_tof_mash_factor()` doesnt exist. Im not 100% sure what it physically is either

KrisThielemans replied:
sure it does. https://github.com/NikEfth/STIR/blob/tof_sino_UCL/src/include/stir/ProjDataInfo.h#L243

TOF mash factor says how many of the "scanner" TOF bins are combined together (to make them larger)

KrisThielemans reviewed (changes requested):
mostly doc changes required, but also an attempt to add TOF resolution in.

if we're renaming `test_consistency_with_root.cxx`, then I vote for `test_consistency_with_GATE.cxx`, or at least `test_consistency_with_ROOT.cxx`

robbietuk committed:
Rename `test_consistency_with_root` to `test_consistency_with_GATE`

Filename and class name change

robbietuk committed:
Class and method documentation and code->method refactoring

robbietuk committed:
[ci skip] CTests save data as CSV with tolerance information.

The first column of the output files is now a key to the row information. 

Also included a file reformat

robbietuk commented:
I electect to add the distance tolerance information into the CTest output file. The first row of the file now indicates the type of information, in a string, that the row corresponds with. The python script has been updated to process this data. Example output for `non_TOF__voxel_data_1.csv`:
```
tolerance,6.66983
original coordinates,0,0,75.3115
0,0,0,75.21
1,0,0,75.21
2,0,0,75.21
3,0,0,75.21
4,0,0,75.21
5,0,0,75.21
6,0,0,75.21
...
```

robbietuk committed:
[ci skip] Change python data extraction to use csv

robbietuk committed:
TOF threshold is a function of TOF resolution. 

This required the earlier initialization of the lm_data_sptr

robbietuk commented:
It resulted in the same `test_time_of_flight` failure, which is expected. I think I addressed all the suggestions in the review.

KrisThielemans commented on examples/ROOT_files/ROOT_STIR_consistency/DebugScripts/debug_consistency_with_root.py:
```diff
@@ -74,27 +76,157 @@ def plot_1D_distance_histogram(distances, n_bins=100, logy=False, pretitle=""):
     """
     Helper class.
-    Loads and converts the data outputs by the `test_view_offset_root` test in STIR into coordinates.
+    Loads and converts the data output by the `test_consistency_with_root` test.
```
```suggestion
    Loads and converts the data output by the `test_consistency_with_GATE` test.
```

KrisThielemans commented on examples/ROOT_files/ROOT_STIR_consistency/DebugScripts/debug_consistency_with_root.py:
```diff
@@ -74,27 +76,157 @@ def plot_1D_distance_histogram(distances, n_bins=100, logy=False, pretitle=""):
-    def __init__(self, filename, tolerance=6.66983):
+    def __init__(self, filename):
         """
         :param filename: Filename of the file output by `test_view_offset_root` (.txt file)
```
```suggestion
        :param filename: Filename of the file output by `test_view_offset_GATE` (.csv file)
```

KrisThielemans commented on examples/ROOT_files/ROOT_STIR_consistency/DebugScripts/debug_consistency_with_root.py:
```diff
@@ -74,27 +76,157 @@ def plot_1D_distance_histogram(distances, n_bins=100, logy=False, pretitle=""):
         """
         :param filename: Filename of the file output by `test_view_offset_root` (.txt file)
-        :param tolerance: l2-norm tolerance that classified a "failed" event. Default = 6.66983, from previous studies.
+        from previous studies.
```
```suggestion
```
truncated?

KrisThielemans commented on examples/ROOT_files/ROOT_STIR_consistency/DebugScripts/debug_consistency_with_root.py:
```diff
@@ -106,46 +238,12 @@ def __init__(self, filename, tolerance=6.66983):
-        if tolerance is not None:
-            print(f"Overwriting tolerance value as {tolerance}")
-            self.tolerance = tolerance
+        print(f"Overwriting tolerance value as {tolerance}")
```
```suggestion
        print(f"Overriding tolerance value as {tolerance}")
```

KrisThielemans commented on examples/ROOT_files/ROOT_STIR_consistency/DebugScripts/debug_consistency_with_root.py:
```diff
@@ -261,29 +357,51 @@ def print_axis_biases():
-print("\nUSAGE: After `make test` or `test_view_offset_root` has been run,\n"
-      "run `debug_view_offset_consistency` from `pretest_output` directory or input that directory as an argument.\n")
+def main():
+    print("\nUSAGE: After `make test` or `test_view_offset_root` has been run,\n"
```
```suggestion
    print("\nUSAGE: After `make test` or `test_view_offset_GATE` has been run,\n"
```

KrisThielemans commented on src/recon_test/test_consistency_with_GATE.cxx:
```diff
@@ -0,0 +1,477 @@
+    Copyright (C) 2017, 2022, UCL
+    This file is part of STIR.
+
+    This file is free software; you can redistribute it and/or modify
+    it under the terms of the GNU Lesser General Public License as published by
+    the Free Software Foundation; either version 2.1 of the License, or
+    (at your option) any later version.
+
+    This file is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU Lesser General Public License for more details.
+    See STIR/LICENSE.txt for details
```
this needs to change to the Apache 2.0 license

KrisThielemans commented on src/recon_test/test_consistency_with_GATE.cxx:
```diff
@@ -0,0 +1,477 @@
+ *
+ * TOF tests
+ * Follows approximately the same procedure as non-TOF tests but, instead of closest approach voxel along the LOR, the voxel
+ * with maximum value is tested. If that voxel exceeds the threshold distance, the test is considered a failer.
```
```suggestion
 * with maximum value is tested. If that voxel exceeds the threshold distance, the test is considered a failure.
```

KrisThielemans commented on src/recon_test/test_consistency_with_GATE.cxx:
```diff
@@ -0,0 +1,477 @@
+   * Main method for running the tests.
+   * For each list mode file, setups up the original point source positions and runs the tests and post processes the results.
+   * Records the pass failure of each file and logs to console.
+   * \returns 0 if all tests passed, 1 otherwise
```
```suggestion
```
`void` function doesn't return anything

KrisThielemans commented on src/recon_test/test_consistency_with_GATE.cxx:
```diff
@@ -0,0 +1,477 @@
+  /*! Test if the voxel with the highest value in the LOR (probabilities) is within TOF_distance_threshold to the original_coords.
+   * If it is, pass with true, otherwise fales.
+   * @param probabilities ProjMatrixElemsForOneBin object of a list mode event
+   * @return True if within TOF_distance_threshold of original_coords, else false.
```
```suggestion
   * @return \c true if within TOF_distance_threshold of original_coords, else \c false.
```

KrisThielemans commented on src/recon_test/test_consistency_with_GATE.cxx:
```diff
@@ -0,0 +1,477 @@
+  num_failed_TOF_lor_events = 0;
+  num_events_tested = 0;
+
+  // Failure conditioner and recording
```
???

KrisThielemans replied:
```suggestion
  // Find threshold for failure
```

KrisThielemans reviewed (changes requested):
some stuff in comments. Ideally batch them all up (and [ci skip]')

KrisThielemans commented on src/recon_test/test_consistency_with_GATE.cxx:
```diff
@@ -0,0 +1,477 @@
+    Copyright (C) 2017, 2022, UCL
+    This file is part of STIR.
+
+    This file is free software; you can redistribute it and/or modify
+    it under the terms of the GNU Lesser General Public License as published by
+    the Free Software Foundation; either version 2.1 of the License, or
+    (at your option) any later version.
+
+    This file is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU Lesser General Public License for more details.
```
```suggestion
    SPDX-License-Identifier: Apache-2.0 AND License-ref-PARAPET-license
```

KrisThielemans committed:
fixes to comments [ci skip]

KrisThielemans commented on examples/ROOT_files/ROOT_STIR_consistency/DebugScripts/debug_consistency_with_root.py:
```diff
@@ -74,27 +76,156 @@ def plot_1D_distance_histogram(distances, n_bins=100, logy=False, pretitle=""):
     """
     Helper class.
-    Loads and converts the data outputs by the `test_view_offset_root` test in STIR into coordinates.
+    Loads and converts the data output by the `test_consistency_with_root` test.
```
```suggestion
    Loads and converts the data output by the `test_consistency_with_GATE` test.
```

KrisThielemans committed:
minor fixes to comments [ci skip]

KrisThielemans reviewed (approved):
all done

robbietuk commented:
Thank you for finalising that for me 😄
